sycl/source/detail/graph_impl.cpp

```diff
@@ -1432,36 +1436,91 @@ void exec_graph_impl::update(
   NeedScheduledUpdate |= MExecutionEvents.size() > 0;
 
   if (NeedScheduledUpdate) {
+    // Copy the list of nodes as we may need to modify it
+    auto NodesCopy = Nodes;
+
+    // If the graph contains host tasks we need special handling here because
+    // their state lives in the graph object itself, so we must do the update
+    // immediately here. Whereas all other command state lives in the backend so
+    // it can be scheduled along with other commands.
+    if (MContainsHostTask) {
+      std::vector<std::shared_ptr<node_impl>> HostTasks;
+      // Remove any nodes that are host tasks and put them in HostTasks
+      auto RemovedIter = std::remove_if(
+          NodesCopy.begin(), NodesCopy.end(),
+          [&HostTasks](const std::shared_ptr<node_impl> &Node) -> bool {
+            if (Node->MNodeType == node_type::host_task) {
+              HostTasks.push_back(Node);
+              return true;
+            }
+            return false;
+          });
+      // Clean up extra elements in NodesCopy after the remove
+      NodesCopy.erase(RemovedIter, NodesCopy.end());
+
+      // Update host-tasks synchronously
+      for (auto &HostTaskNode : HostTasks) {
+        updateImpl(HostTaskNode);
+      }
+    }
+
     auto AllocaQueue = std::make_shared<sycl::detail::queue_impl>(
         sycl::detail::getSyclObjImpl(MGraphImpl->getDevice()),
         sycl::detail::getSyclObjImpl(MGraphImpl->getContext()),
         sycl::async_handler{}, sycl::property_list{});
-    // Don't need to care about the return event here because it is synchronous
```

Comment:
Could we replace this with a comment about why we do need to care about the return even for a synchronous UR call.

Reply:
Done!